### ISSUE TITLE
Unsound interaction between first-class modules and polymorphic methods

### ISSUE THREAD

**Issue by yallop** [ocaml/ocaml#12910]
There's a bug in the type checker that treats equal module types as incompatible.  ObSegFault:

```
$ cat bang.ml
```
```ocaml
module F(V: sig
             module type X = sig type t end
             module type Y = sig type t end
             val x_eq_y : (<m: 'a.(module X) -> 'a>,
                           <m: 'a.(module Y) -> 'a>) Type.eq
           end) =
struct
  let f : ((<m: 'a.(module V.X) -> 'a>,
            <m: 'a.(module V.Y) -> 'a>) Type.eq, string) Either.t -> unit =
    function Right s -> print_endline s
           | _ -> .

  let () = f (Left V.x_eq_y)
end

module M = struct module type X = sig type t end
                  module type Y = X
                  let x_eq_y = Type.Equal end
include F(M)
```
```
$ ocaml bang.ml 
Segmentation fault
```

As a malus, there's something wrong with the treatment of subtyping coercions involving polymorphic methods and first-class modules, too.  This program is accepted:

```ocaml
module type S = sig        end
module type T = sig type t end

type s = < m: 'a. (module S) -> 'a >
type t = < m: 'a. (module T) -> 'a >

let f (s : s) = (s : s :> t)
```

but this slight (and rather innocent-looking) variant is rejected:

```ocaml
module type S = sig type t end
module type T = sig type t end

type s = < m: 'a. (module S) -> 'a >
type t = < m: 'a. (module T) -> 'a >

let f (s : s) = (s : s :> t)
```

with the following message:

```
7 | let f (s : s) = (s : s :> t)
                    ^^^^^^^^^^^^
Error: Type "s" = "< m : 'a. (module S) -> 'a >" is not a subtype of
         "t" = "< m : 'a. (module T) -> 'a >"
       Type "'a" is not a subtype of "'a0"
```

**Comment by stedolan** [comment]
The underlying problem here is that the types `<m: 'a. (module X) -> 'a>` and `<m: 'a. (module Y) -> 'a>` do not unify:
```
# let f (x : <m : 'a . (module X) -> 'a>) = (x : <m : 'a . (module Y) -> 'a>);;
Error: This expression has type < m : 'a. (module Y) -> 'a >
       but an expression was expected of type < m : 'a. (module Y) -> 'a >
       Types for method m are incompatible
```
The failure to unify here is what allows the typechecker to refute the `Left Type.Equal` case in the example.

The failure is because of misuse of global mutable state in the typechecker. When comparing the two polymorphic types, the typechecker (in `enter_poly`) initialises `univar_pairs`, so that the two universal variables `'a` in the function return types will be unifiable.

However, before it gets there it calls `package_subtype` to compare the two module types, which calls into the module system, which eventually calls `Includecore.type_declarations` (because the two module types both contain a `type t`). To compare the two types, `Includecore` uses `Ctype.equal`, which calls `Ctype.eqtype_list`. Unfortunately, that function does this:

```
(* Must empty univar_pairs first *)
let eqtype_list rename type_pairs subst env tl1 tl2 =
  univar_pairs := [];
  let snap = Btype.snapshot () in
  Misc.try_finally
    ~always:(fun () -> backtrack snap)
    (fun () -> eqtype_list rename type_pairs subst env tl1 tl2)
```
So `univar_pairs` is clobbered during the nested use of the typechecker machinery when comparing package types, so that it does not have the right contents when the `'a` variables are compared.

This specific issue can probably be fixed by ensuring that `eqtype_list` preserves the original value of `univar_pairs` rather than overwriting it. However, it makes me wonder whether there are other bugs of this form involving global mutable state being overwritten during `package_subtype`.

**Comment by yallop** [comment]
Fixed by #12920 (88be73e3bf26838a3716fdeeabf6fa212f480117)
